### ISSUE TITLE
Implement miTCCR

### ISSUE THREAD

**Issue by sinui0** [privacy-ethereum/mpz#7]
Rereading https://eprint.iacr.org/2019/1168.pdf it occurs to me that we never implemented the recommendation to randomly sample the initial gate id while garbling a circuit. We should do this to achieve multi-instance TCCR.

**Comment by themighty1** [comment]
My understanding is that we don't need to worry about miTCCR, here's why:
Edit: wrong understang, see newer comment below

~~GKWY19 (https://eprint.iacr.org/2019/ analyzes half-gates's notion  "circular correlation robustness for naturally derived keys” in Section 5 and says that this notion "can be viewed as a form of tweakable circular correlation robustness (tccr)".~~
~~Then it says that tccr for half-gates "is overkill, and we show now that a family of hash functions satisfying the notion can be constructed from any H that is ccr".~~
~~Thus, we don't need tweakable ccr but just ccr. Since it is the re-use of a tweak that is the premise of the miTCCR paper, using ccr instead of tccr means the attack from the paper does not apply aymore.~~

~~I will double check about this line of reasoning with the emp-tool authors.~~

**Comment by themighty1** [comment]
I realized that the emp-tool author hasn't responded to my q about this matter: https://github.com/emp-toolkit/emp-tool/issues/179
EDIT: he did respond

I think we should play it safe and implement the random sampling

**Comment by themighty1** [comment]
There are 3 things I learned from the "multi-instance (MI)" paper (https://eprint.iacr.org/2019/1168)

1. There is an attack on GKWY19's TCCR hash where security degrades with the amount of AND gates garbled (even across multiple independently garbled circuits), quoting

>  They [GKWY19] did not focus on obtaining better concrete security, and indeed, in Appendix C we show that using their hash function [TCCR] in the half-gates scheme would admit an attack similar to the one described above.

2. Then the paper shows a more secure miTCCR hash.

3. Then they show that miTCCR hash's security is tight only when 1 circuit is garbled.When multiple circuits are garbled, they suggest the tweak to randomize gate ids.


In light of this, we should both:
- replace TCCR with miTCCR
- add gate id randomization

**Comment by themighty1** [comment]
miTCCR explained by one of the authors: https://www.youtube.com/watch?v=MKGqA0K9d8M

**Comment by themighty1** [comment]
We need to examine how findings of this paper https://eprint.iacr.org/2024/163 apply to us.

**Comment by themighty1** [comment]
The latest paper 2024/163 proves some security bounds when using miTCCR hash with a leaky delta (like the delta in KOS).
In the leaky-delta setting miTCCR security is worse than in the original miTCCR paper 2019/1168, but I couldn't exactly figure out how much worse it is.

Regardless of the 2024/163 findings, we still need to

 - replace TCCR with miTCCR 
 - add gate id randomization

**Comment by themighty1** [comment]
The attack on the TCCR hash shown in the miTCCR paper also applies to multiple circuits with different deltas. To quote:

![ee](https://github.com/user-attachments/assets/cf6d7a6f-d97c-4b8a-ba78-b28c1a9c4f21)

the "shift R" is the delta.

**Comment by themighty1** [comment]
Relevant: https://eprint.iacr.org/2025/1818

Also claims to fix a flaw in the security proof of the Ferret paper.

**Comment by themighty1** [comment]
Comparing performance of miTCCR hash to the current TCCR hash:
- native performance with AES-NI: miTCCR is ~3,5x slower
- wasm performance: miTCCR is ~2.5x slower

miTCCR was implemented and benched on this branch https://github.com/themighty1/mpz/blob/46ee34d44e06c58a045c7573f5260aee059e05ed/crates/core/src/aes.rs#L40

**Comment by heeckhau** [comment]
@themighty1 Can this  be close?

**Comment by themighty1** [comment]
nope, this should be kept open, we only randomized gate id, still need to `replace TCCR with miTCCR`
